packages/client-utils/src/vstorage-client.js

```diff
@@ -0,0 +1,423 @@
+  'No valid RPC address found in config';
+
+const MINIMUM_HEIGHT = 1n;
+
+export const PATH_PREFIX = '/agoric.vstorage.Query';
+
+export const PATHS = /** @type {const} */ ({
+  CHILDREN: 'Children',
+  DATA: 'Data',
+});
+
+const codecs = /** @type {const} */ ({
+  [PATHS.CHILDREN]: {
+    request: QueryChildrenRequest,
+    response: QueryChildrenResponse,
+  },
+  [PATHS.DATA]: {
+    request: QueryDataRequest,
+    response: QueryDataResponse,
+  },
+});
+
+/**
+ * @param {ReturnType<typeof parseValue>} value
+ */
+const assertValueStructure = value => {
+  if (!isStreamCell(value))
+    throw Error(`Expected a stream cell, got ${quote(value)}`);
+};
+
+/**
+ * @param {ReturnType<parseValue>} cell
+ */
+const isDataString = cell => !!cell && typeof cell === 'string';
+
+/**
+ * @param {{
+ *  errorMessage: string;
+ *  height?: Height;
+ *  kind: PATHS[keyof PATHS];
+ *  path: string;
+ *  rpcAddress?: MinimalNetworkConfig['rpcAddrs'][0];
+ * }} options
+ */
+const makeVstorageError = ({ errorMessage, height, kind, path, rpcAddress }) =>
+  makeError(
+    `Cannot read '${kind}' of '${path}' ${
+      rpcAddress ? `from '${rpcAddress}'` : ''
+    } ${height ? `at height '${height}'` : ''} due to error: ${errorMessage}`,
+  );
+
+/**
+ * @template {any} T
+ * @param {{queryClient: ReturnType<typeof makeQueryClient>}} powers
+ * @param {string} path
+ * @param {Partial<{ compat: boolean }>} [options]
+ * @returns {Topic<T>}
+ */
+const makeBlockTopic = ({ queryClient }, path, options) => ({
+  latest: async height => {
+    if (height && height < MINIMUM_HEIGHT)
+      throw makeVstorageError({
+        errorMessage: `${INVALID_HEIGHT_ERROR_MESSAGE} ${height}`,
+        height,
+        kind: PATHS.DATA,
+        path,
+      });
+
+    const response = await queryClient.queryAbci(path, {
+      height: height && height > MINIMUM_HEIGHT ? height - 1n : height,
+      kind: PATHS.DATA,
+    });
+    const parsedValue = parseValue(response.value);
+
+    if (isDataString(parsedValue))
+      return /** @type {Update<T>} */ ({
+        blockHeight: height,
+        value: parsedValue,
+      });
+
+    if (options?.compat) {
+      assertValueStructure(parsedValue);
+      const data = /** @type {StreamCell} */ (parsedValue);
+      const blockHeight = BigInt(data.blockHeight);
+      return /** @type {Update<T>} */ ({
+        blockHeight,
+        value: data.values.reverse().find(Boolean),
+      });
+    }
+
+    throw Error(`Expected a data cell, got ${quote(parsedValue)}`);
+  },
+});
+
+/**
+ * @param {{fetch: typeof window.fetch}} powers
+ * @param {MinimalNetworkConfig} config
+ */
+const makeQueryClient = ({ fetch }, config) => {
+  if (!config.rpcAddrs.length)
+    return Fail`${INVALID_RPC_ADDRESS_ERROR_MESSAGE} ${quote(config)}`;
+
+  /**
+   * @param {Height} height
+   * @param {PATHS[keyof PATHS]} kind
+   * @param {string} path
+   */
+  const computeVstorageAbciRoute = (height, kind, path) =>
```

Comment:
Is "Route" sufficiently clear?
```suggestion
  const toVstorageAbciQueryUrlPath = (height, kind, path) =>
```
or maybe better still,
```js
  const toVstorageAbciQueryUrl = (baseUrl, height, kind, path) => {
    const hexData = encodeHex(codecs[kind].request.toProto({ path }));
    const urlPath = encodeURI(
      `/abci_query?data=0x${hexData}&height=${height}&path="${PATH_PREFIX}/${kind}"`,
    );
    return baseUrl + urlPath;
  };
```